flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerBaseTest.java

```diff
@@ -88,195 +87,296 @@ public void testKeyValueDeserializersSetIfMissing() throws Exception {
+	 * we set a timeout because the test will not finish if the logic is broken
+	 */
+	@Test(timeout=5000)
+	public void testDoesNotWaitForPendingRecordsIfFlushingDisabled() throws Throwable {
 
-	// ------------------------------------------------------------------------
+		KafkaProducer mockProducer = mock(KafkaProducer.class);
+		final DummyFlinkKafkaProducer<String> producer = new DummyFlinkKafkaProducer<>(
+			FakeStandardProducerConfig.get(), null, mockProducer);
+		producer.setFlushOnCheckpoint(false);
 
-	private static class DummyFlinkKafkaProducer<T> extends FlinkKafkaProducerBase<T> {
-		private static final long serialVersionUID = 1L;
+		final OneInputStreamOperatorTestHarness<String, Object> testHarness =
+			new OneInputStreamOperatorTestHarness<>(new StreamSink(producer));
 
-		private transient MockProducer prod;
-		private AtomicBoolean snapshottingFinished;
+		testHarness.open();
 
-		@SuppressWarnings("unchecked")
-		public DummyFlinkKafkaProducer(Properties producerConfig, KafkaPartitioner partitioner, AtomicBoolean snapshottingFinished) {
-			super("dummy-topic", (KeyedSerializationSchema< T >) mock(KeyedSerializationSchema.class), producerConfig, partitioner);
-			this.snapshottingFinished = snapshottingFinished;
-		}
+		testHarness.processElement(new StreamRecord<>("msg"));
 
-		// constructor variant for test irrelated to snapshotting
-		@SuppressWarnings("unchecked")
-		public DummyFlinkKafkaProducer(Properties producerConfig, KafkaPartitioner partitioner) {
-			super("dummy-topic", (KeyedSerializationSchema< T >) mock(KeyedSerializationSchema.class), producerConfig, partitioner);
-			this.snapshottingFinished = new AtomicBoolean(true);
-		}
+		// make sure that all callbacks have not been completed
+		verify(mockProducer, times(1)).send(any(ProducerRecord.class), any(Callback.class));
 
-		@Override
-		protected <K, V> KafkaProducer<K, V> getKafkaProducer(Properties props) {
-			this.prod = new MockProducer();
-			return this.prod;
-		}
+		// should return even if there are pending records
+		testHarness.snapshot(123L, 123L);
 
-		@Override
-		public void snapshotState(FunctionSnapshotContext ctx) throws Exception {
-			// call the actual snapshot state
-			super.snapshotState(ctx);
-			// notify test that snapshotting has been done
-			snapshottingFinished.set(true);
-		}
+		testHarness.close();
+	}
 
-		@Override
-		protected void flush() {
-			this.prod.flush();
-		}
+	// ------------------------------------------------------------------------
 
-		public MockProducer getProducerInstance() {
-			return this.prod;
-		}
-	}
+	private static class DummyFlinkKafkaProducer<T> extends FlinkKafkaProducerBase<T> {
+		private static final long serialVersionUID = 1L;
+		
+		private final static String DUMMY_TOPIC = "dummy-topic";
 
-	private static class MockProducer<K, V> extends KafkaProducer<K, V> {
-		List<Callback> pendingCallbacks = new ArrayList<>();
+		private final KafkaProducer mockProducer;
```

Comment:
Will change to `transient`.

Curious, though: is this just for documentation purposes for readers of the code?

Otherwise, we need the `mockProducer` throughout the whole life cycle of the object (I tried instantiating a mocked `KafkaProducer` in `open()` like what we would normally do for the non-serializable fields, but it won't work because of how the `testPartitionerOpenedWithDeterminatePartitionList` needs to mock more method calls before calling `open()`). Also in the tests we're not serializing the instances anyway.